storage/innobase/xtrabackup/src/ds_fifo.cc

```diff
@@ -0,0 +1,167 @@
+/******************************************************
+Copyright (c) 2023 Percona LLC and/or its affiliates.
+
+FIFO datasink implementation for XtraBackup.
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; version 2 of the License.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
+
+*******************************************************/
+
+#include <my_base.h>
+#include <my_io.h>
+#include <my_thread_local.h>
+#include <mysql/service_mysql_alloc.h>
+#include <mysys_err.h>
+#include <map>
+#include "common.h"
+#include "datasink.h"
+#include "file_utils.h"
+#include "msg.h"
+
+typedef struct {
+  /* List of FIFO files to be used for stream */
+  std::map<std::string, File> FIFO_list;
+  /* Mutex protecting FIFO list */
+  pthread_mutex_t fifo_mutex;
+} ds_fifo_ctxt_t;
+
+typedef struct {
+  File fd;
+  char *path;
+  ds_fifo_ctxt_t *fifo_context;
+} ds_fifo_file_t;
+
+extern uint xtrabackup_parallel_stream;
+extern uint xtrabackup_stream_timeout;
+static ds_ctxt_t *fifo_init(const char *root);
+static ds_file_t *fifo_open(ds_ctxt_t *ctxt, const char *path, MY_STAT *mystat);
+static int fifo_write(ds_file_t *file, const void *buf, size_t len);
+static int fifo_close(ds_file_t *file);
+static void fifo_deinit(ds_ctxt_t *ctxt);
+
+datasink_t datasink_fifo = {&fifo_init, &fifo_open,  &fifo_write,
+                            nullptr,    &fifo_close, &fifo_deinit};
+
+static ds_ctxt_t *fifo_init(const char *root) {
+  ds_fifo_ctxt_t *fifo_context = new ds_fifo_ctxt_t;
+  File fd;
+  ds_ctxt_t *ctxt = new ds_ctxt_t;
+  char fullpath[FN_REFLEN];
+
+  if (pthread_mutex_init(&fifo_context->fifo_mutex, NULL)) {
+    msg("fifo_init: pthread_mutex_init() failed.\n");
+    return NULL;
+  }
+
+  if (my_mkdir(root, 0777, MYF(0)) < 0 && my_errno() != EEXIST &&
+      my_errno() != EISDIR) {
+    char errbuf[MYSYS_STRERROR_SIZE];
+    my_error(EE_CANT_MKDIR, MYF(ME_BELL), root, my_errno(),
+             my_strerror(errbuf, sizeof(errbuf), my_errno()));
+    return NULL;
+  }
+
+  for (uint i = 0; i < xtrabackup_parallel_stream; i++) {
+    std::string path = "thread_" + std::to_string(i);
+    fn_format(fullpath, path.c_str(), root, "", MYF(MY_RELATIVE_PATH));
+    mkfifo(fullpath, 0666);
```

Comment:
Do we delete this FIFO file at the end of the backup?

Reply:
> Handle error in case mkfifo FAIL.

Done

> Do we delete this FIFO file at the end of the backup?

The reader side does it (xbcloud put or xbstream)